src/public/logger.h

```diff
@@ -0,0 +1,46 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef NINJA_PUBLIC_LOGGER_H_
+#define NINJA_PUBLIC_LOGGER_H_
+
+#include <iostream>
+#include <string>
+
+namespace ninja {
+
+class Logger {
```

Comment:
This opens the question on whether to support compile-time polymorphism (templates) or run-time polymorphism (virtual member functions).

It's clear to me that the current codebase of Ninja is very runtime oriented on things like interfaces, and converting large parts of the code to templates would meet a lot of resistance.

Of course, compile-time polymorphism can be (depending on implementation details) strictly more flexible than run-time polymorphism, as you can pass an abstract base class as the type to instantiate the template with.